src/normalizer.ts

```diff
@@ -76,325 +106,192 @@ function normalizeV0Schema(parsed: parser.V0Schema): XtpSchema {
+      return new ArrayType(this.annotateType(s.items)!, s)
     }
 
-    if (im.output?.$ref) {
-      const path = `#/imports/${name}/output`
+    // if additionalProperties is present it's a map
+    if (s.additionalProperties) {
+      // TODO remove this legacy code
+      // we need to derive old type here
+      s.type = 'map'
 
-      normalizeProp(
-        normIm.output!,
-        querySchemaRef(schemas, im.output.$ref, path),
-        path
-      )
+      return new MapType(this.annotateType(s.additionalProperties)!, s)
     }
-    if (im.output?.items?.$ref) {
-      const path = `#/imports/${name}/output/items`
-
-      normalizeProp(
-        normIm.output!.items!,
-        querySchemaRef(schemas, im.output.items.$ref, path),
-        path
-      )
-    }
-
-    validateArrayItems(normIm.input?.items, `#/imports/${name}/input/items`);
-    validateArrayItems(normIm.output?.items, `#/imports/${name}/output/items`);
 
-    imports.push(normIm)
-  }
+    switch (s.type) {
+      case 'string':
+        if (s.format === 'date-time') return new DateTimeType(s)
+        return new StringType(s)
+      case 'integer':
+        return new Int32Type(s)
+      case 'boolean':
+        return new BooleanType(s)
+      case 'buffer':
+        return new BufferType(s)
+      case 'number':
+        if (s.format === 'int32') return new Int32Type(s)
+        if (s.format === 'int64') return new Int64Type(s)
+        if (s.format === 'float') return new FloatType(s)
+        if (s.format === 'double') return new DoubleType(s)
+        throw new Error(`IDK how to parse this number ${JSON.stringify(s)}`)
+    }
 
-  for (const name in schemas) {
-    const schema = schemas[name]
-    const error = detectCircularReference(schema);
-    if (error) {
-      throw error;
+    // if we get this far, we don't know what
+    // this node is let's just keep drilling down
+    for (const key in s) {
+      if (Object.prototype.hasOwnProperty.call(s, key)) {
+        const child = s[key]
+        if (child && typeof child === 'object' && !Array.isArray(child)) {
+          const t = this.annotateType(child);
+          if (t) child.xtpType = t
+        }
+      }
     }
   }
+}
 
-  return {
-    version,
-    exports,
-    imports,
-    schemas,
-  }
+function normalizeV1Schema(parsed: parser.V1Schema): XtpSchema {
+  const normalizer = new V1SchemaNormalizer(parsed)
+  return normalizer.normalize()
 }
 
 export function parseAndNormalizeJson(encoded: string): XtpSchema {
-  const parsed = parser.parseJson(encoded)
+  const { doc, errors } = parser.parseAny(JSON.parse(encoded))
 
-  if (parser.isV0Schema(parsed)) {
-    return normalizeV0Schema(parsed)
-  } else if (parser.isV1Schema(parsed)) {
-    return normalizeV1Schema(parsed)
-  } else {
-    throw new ValidationError("Could not normalize unknown version of schema", "#");
-  }
-}
-
-function detectCircularReference(schema: Schema, visited: Set<string> = new Set()): ValidationError | null {
```

Comment:
I deleted this but have not fixed this yet

Reply:
I don't think we need to block circular refs just get clever about how to process it